blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,177 @@
+---
+title: What features PCIe has?
+abstract: 'Introductory blog post to PCIe features.
+          In this article you can read what PCIe capability is
+          and see examples of such capabilities.'
+cover: /covers/PCIExpress.jpg
+author: marek.kasiewicz
+layout: post
+published: true
+date: 2020-08-05
+archives: "2020"
+
+tags:
+  - Firmware
+categories:
+  - Firmware
+  - IoT
+  - Miscellaneous
+  - OS Dev
+  - App Dev
+  - Security
+  - Manufacturing
+
+---
+
+## What this blog is about
+
+Goal of this post is to introduce readers to the concept of PCIe capabilities.
+It briefly describes what capability is and how it works. It also lists some
+example capability described in PCIe specification.
+
+## Introduction
+
+Every device in the PCIe tree has at least one function. Each function has its
+configuration space. Configuration space is device registers mapped to either
+I/O memory space (used mostly by PCI devices) or to normal memory space. PCI
+capable configuration space has 256 bytes, extended configuration space used in
+PCIe has 4096 bytes.
+
+In this memory we can find most importantly 2 structures.
+
+- **Header**  
+![config header](/img/Pci-config-space.png)  
+It contains information like vendor ID, and device ID, pointers to memory
+assigned to the function and pointer to first
+capability in linked list.
+- **Linked list of capabilities**  
+Each capability has its own header, it contains ID of the
+capability, its version and pointer to the next capability in the list. Beneath
+this header are mapped registers needed by capabilities.
+
+To check if a given capability is implemented by function, a software has to
+search through the list and check if a given capability ID is present in it.
+
+## What PCIe capability actually is
+
+It is nothing more than a predefined feature of the device's function, a feature
+that is known to be possible to be implemented in all devices, but most of them
+are optional. Every capability has registers in a device mapped to configuration
+space. Those registers are a kind of interface to the capability. Some are read
+only and contain information needed by the software, others are read/write and
+can be used to pass some information to the function.
+
+## List of capabilities structures:
+(This list can be incomplete)
+
+- **PCI Express Capability register block**
+
+This is one of the most important capability structures, it must be present in
+all PCIe functions. It contains information about:
+
+1. Device e.g.:  
+Maximum supported payload size, maximum accepted latency
```

Comment:
Short explanation what payload are we talking about and what the latency refers to